ggshield/cmd/hmsl/fingerprint.py

```diff
@@ -0,0 +1,174 @@
+    key: Optional[str]
+    value: str
+
+
+# Methods to compute names for secrets
+# They are useful to help the user identify the secret that may leaked
+# with a more "human-readable" string than a hash.
+# Takes the secret and optional key as input and returns a string.
+NamingStrategy = Callable[[SecretWithKey], str]
+
+NAMING_STRATEGIES: Dict[str, NamingStrategy] = {
+    "censored": lambda secret: censor_string(secret.value),
+    "cleartext": lambda secret: secret.value,
+    "none": lambda _: "",
+    "key": lambda secret: secret.key or censor_string(secret.value),
+}
+
+
+# Command
+
+input_arg = click.argument(
+    "path",
+    type=click.Path(
+        exists=True,
+        dir_okay=False,
+        readable=True,
+        allow_dash=True,
+    ),
+)
+
+naming_strategy_option = click.option(
+    "--naming-strategy",
+    "-n",
+    type=click.Choice(list(NAMING_STRATEGIES.keys())),
+    default="key",
+    show_default=True,
+    help="""Strategy to generate the hints.
+            With "censored", only the firsts and last characters are displayed.
+            With "cleartext", the full secret is used as a hint (Not recommended!).
+            With "none", no hint is generated.
+            With "key", the key name is selected if available (e.g. in .env files), otherwise censored is used.""",
+    callback=lambda _, __, value: NAMING_STRATEGIES[value],
+)
+
+full_hashes_option = click.option(
+    "-f",
+    "--full-hashes",
+    is_flag=True,
+    default=False,
+    help="Send full hashes instead of prefixes.",
+)
+
+
+def validate_prefix(prefix: str) -> str:
+    if prefix != "" and not prefix.endswith("-"):
+        prefix = prefix + "-"
+    return prefix
+
+
+@click.command()
+@add_common_options()
+@click.option(
+    "--prefix",
+    "-p",
+    default="",
+    help="Prefix for output file names.",
+    callback=lambda _, __, value: validate_prefix(value),
+)
+@full_hashes_option
+@naming_strategy_option
+@input_arg
+def fingerprint_cmd(
+    path: str, prefix: str, naming_strategy: NamingStrategy, full_hashes: bool, **_: Any
+) -> int:
+    """
+    Collect secrets and prepare them to be queried.
+    """
+    # Opens the file or stdin
+    input = cast(TextIO, click.open_file(path, "r"))
+
+    # Prepare and write the output files
+    secrets = list(collect(input))
+    result = prepare(secrets, naming_strategy, full_hashes=full_hashes)
+    write_outputs(result, prefix)
+
+    display_info(f"Prepared {len(secrets)} secrets.")
+    return 0
+
+
+# Helper methods
+
+
+def collect(
+    input: TextIO, source: SourceType = SourceType.FILE
+) -> Iterator[SecretWithKey]:
+    """
+    Collect the secrets from the source
+    """
+    for line in input:
+        secret = line.strip()
+        if secret == "":
+            # Skip empty lines
+            continue
+        if source == SourceType.FILE:
+            yield SecretWithKey(value=secret, key=None)
+
+
+def prepare(
+    secrets: List[SecretWithKey],
```

Comment:
unless your function needs random access, it's better to make arguments take `Iterable[T]` instead of `List[T]` because it makes it possible for callers to pass generators.